zerver/views/message_send.py

```diff
@@ -339,3 +348,77 @@ def render_message_backend(
 
     rendering_result = render_markdown(message, content, realm=user_profile.realm)
     return json_success(request, data={"rendered": rendering_result.rendered_content})
+
+
+REPORT_MESSAGE_REASONS = {
+    "spam": f"{_('Spam')}",
+    "harassment": f"{_('Harassment')}",
+    "inappropriate": f"{_('Inappropriate content')}",
+    "norms": f"{_('Violates community norms')}",
+    "other": f"{_('Other reason')}",
+}
+
+# must be kept in sync with ./static/shared/js/fenced_code.js
+def get_unused_fence(content: str) -> str:
+    # in python, it's one line of code get a uniquely long fence
+    return "`" * (1 + max([2] + [len(x) for x in re.findall(r"`+", content)]))
+
+
+def get_quoted_text(content: str) -> str:
+    fence = get_unused_fence(content)
+    return f"\n{fence}quote\n{content}\n{fence}\n"
+
+
+@has_request_variables
+def report_message_backend(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    message_id: int = REQ(converter=to_non_negative_int, path_only=True),
+    reason: str = REQ(
+        default="other", str_validator=check_string_in(list(REPORT_MESSAGE_REASONS.keys()))
+    ),
+    explanation: Optional[str] = REQ(default=None),
+) -> HttpResponse:
+    request_notes = RequestNotes.get_notes(request)
+    assert request_notes.log_data is not None
+    orig_message, ignored = access_message(user_profile, message_id, lock_message=False)
+    reason_str = REPORT_MESSAGE_REASONS[reason]
+    explanation = "" if explanation is None else explanation.strip()
+    if reason == "other" and explanation == "":
+        raise JsonableError(_("For reason=other, an explanation must be provided."))
+
+    if explanation == "":
+        explanation_str = ".\n\n"
+    else:
+        explanation_str = f" {_('with these notes')}:\n{get_quoted_text(explanation)}\n"
+
+    # put original message at the end, so if rptmsg gets truncated, it's from that
+    if orig_message.recipient.type == Recipient.STREAM:
+        stream = get_stream_by_id_in_realm(orig_message.recipient.type_id, user_profile.realm)
+        recipient = f"#**{stream.name}>{orig_message.topic_name()}**\n"
+    elif orig_message.recipient.type == Recipient.PERSONAL:
+        receiver = get_user_profile_by_id_in_realm(
+            orig_message.recipient.type_id, user_profile.realm
+        )
+        recipient = f"@**{receiver.full_name}|{receiver.id}**\n"
+    else:
+        recipient = " ".join([f"@**|{id}**" for id in get_huddle_user_ids(orig_message.recipient)])
+
+    # single template to make it easy to change later (see
+    user_name, user_id = user_profile.full_name, user_profile.id
+    sender_name, sender_id = orig_message.sender.full_name, orig_message.sender.id
+    quoted_message = get_quoted_text(orig_message.content)
+    rptmsg = f"""@_**{user_name}|{user_id}** reported a message sent by @_**{sender_name}|{sender_id}** {_('to')} {recipient} {_('as containing')} **`{reason_str}`**{explanation_str}
+{_('Original message')}:
+{quoted_message}"""
```

Comment:
Can’t i18n sentence fragments.

This should be a single template that gets translated and then interpolated.

Reply:
DONE.